standalone.go

```diff
@@ -53,29 +64,126 @@ func (s *standalone) pick() int {
 	return rand.IntN(len(s.replicas))
 }
 
+func (s *standalone) handleRedirect(ctx context.Context, cmd Completed, result ValkeyResult) ValkeyResult {
+	if !s.enableRedirect {
+		return result
+	}
+
+	if ret, yes := IsValkeyErr(result.Error()); yes {
+		if addr, ok := ret.IsRedirect(); ok {
+			// Update the latest redirect address before singleflight
+			s.lastRedirectAddr = addr
+			
+			// Use singleflight to ensure only one redirect operation happens at a time
+			// Only proceed if the address matches the latest seen redirect address
+			if err := s.redirectCall.Do(ctx, func() error {
+				// Double-check that this is still the latest redirect address
+				if s.lastRedirectAddr == addr {
+					return s.redirectToPrimary(addr)
+				}
+				// If address doesn't match, another redirect happened, skip this one
+				return nil
+			}); err != nil {
+				// If redirect fails, return the original result
+				return result
+			}
+
+			// Execute the command on the updated primary
+			return s.primary.Do(ctx, cmd)
+		}
+	}
+
+	return result
+}
+
+func (s *standalone) redirectToPrimary(addr string) error {
+	// Create a new connection to the redirect address
+	redirectOpt := *s.opt
+	redirectOpt.InitAddress = []string{addr}
+	redirectConn := s.connFn(addr, &redirectOpt)
+	if err := redirectConn.Dial(); err != nil {
+		return err
+	}
+
+	// Create a new primary client with the redirect connection
+	newPrimary := newSingleClientWithConn(redirectConn, cmds.NewBuilder(cmds.NoSlot), !s.opt.DisableRetry, s.opt.DisableCache, s.retryer, false)
+
+	// Close the old primary and swap to the new one
+	oldPrimary := s.primary
+	s.primary = newPrimary
+	oldPrimary.Close()
+
+	return nil
+}
+
 func (s *standalone) Do(ctx context.Context, cmd Completed) (resp ValkeyResult) {
-	if s.toReplicas(cmd) {
-		return s.replicas[s.pick()].Do(ctx, cmd)
+	attempts := 1
+retry:
+	if s.toReplicas != nil && s.toReplicas(cmd) {
+		resp = s.replicas[s.pick()].Do(ctx, cmd)
+	} else {
+		resp = s.primary.Do(ctx, cmd)
+	}
+
+	// Handle redirects with retry until context deadline
+	resp = s.handleRedirect(ctx, cmd, resp)
+	if ret, yes := IsValkeyErr(resp.Error()); yes {
+		if _, ok := ret.IsRedirect(); ok && s.enableRedirect {
+			// Use retryHandler to handle multiple redirects with context deadline
+			if s.retryer.WaitOrSkipRetry(ctx, attempts, cmd, resp.Error()) {
+				attempts++
+				goto retry
+			}
+		}
```

Comment:
Since we now have the retry structure, I think we don't need `handleRedirect` anymore, or we can largely simplify  `handleRedirect`. The retry structure should look like this:

```suggestion
    if s.enableRedirect {
      if ret, yes := IsValkeyErr(resp.Error()); yes {
		if addr, ok := ret.IsRedirect(); ok {
		    err := s.redirectCall(ctx, func() {return s.redirectToPrimary})
			// Use retryHandler to handle multiple redirects with context deadline
			if err == nil || s.retryer.WaitOrSkipRetry(ctx, attempts, cmd, resp.Error()) {
				attempts++
				goto retry
			}
		}
```